api/admin/controller.py

```diff
@@ -815,3 +823,181 @@ def result_to_row(result):
+            library.name = name
+        if short_name:
+            library.short_name = short_name
+        if registry_short_name:
+            library.library_registry_short_name = registry_short_name
+        if registry_shared_secret:
+            library.library_registry_shared_secret = registry_shared_secret
+
+        if is_new:
+            return Response(unicode(_("Success")), 201)
+        else:
+            return Response(unicode(_("Success")), 200)
+
+    def collections(self):
+        if flask.request.method == 'GET':
+            collections = [
+                dict(
+                    name=collection.name,
+                    protocol=collection.protocol,
+                    libraries=[library.short_name for library in collection.libraries],
+                    external_account_id=collection.external_account_id,
+                    url=collection.url,
+                    username=collection.username,
+                    password=collection.password,
+                    settings={setting.key: setting.value for setting in collection.settings}
+                )
+                for collection in self._db.query(Collection).order_by(Collection.name).all()
+            ]
+
+            return dict(
+                collections=collections,
+                protocols=Collection.PROTOCOLS,
+            )
+
+
+        name = flask.request.form.get("name")
+        if not name:
+            return MISSING_COLLECTION_NAME
+        protocol = flask.request.form.get("protocol")
+        external_account_id = flask.request.form.get("external_account_id")
+        url = flask.request.form.get("url")
+        username = flask.request.form.get("username")
+        password = flask.request.form.get("password")
+        settings = {}
+        if flask.request.form.get("settings"):
+            settings = json.loads(flask.request.form.get("settings"))
+        libraries = []
+        if flask.request.form.get("libraries"):
+            libraries = json.loads(flask.request.form.get("libraries"))
+
+        is_new = False
+
+        collection = get_one(self._db, Collection, name=name)
+        if not collection:
+            if protocol:
+                collection, is_new = get_one_or_create(
+                    self._db, Collection, name=name, protocol=protocol
+                )
+            else:
+                return NO_PROTOCOL_FOR_NEW_COLLECTION
+
+        if protocol:
+            collection.protocol = protocol
+        if external_account_id:
+            collection.external_account_id = external_account_id
+        if url:
+            collection.url = url
+        if username:
+            collection.username = username
+        if password:
+            collection.password = password
+
+        for key, value in settings.items():
+            collection.setting(key).value = value
+        for setting in collection.settings:
+            if setting.key not in settings.keys():
+                self._db.delete(setting)
```

Comment:
I'm a little worried about this bit. I don't want the client to be able to set arbitrary settings. Each protocol has a limited number of settings that can be set, e.g. "website_id" for Overdrive. I'd prefer if only settings associated with collection.protocol could be set here, and all others had their values removed.

This might require changing core to add an inventory of allowable settings to each CirculationAPI subclass, so that it's kept in the same place as the code that implements the protocol.

Another problem with this: if you change the protocol, say from Bibliotheca to Overdrive, then a previously invalid setting becomes required. I think right now you won't be able to change the protocol from non-Overdrive to Overdrive. This needs to work somehow. Maybe you can change the protocol, put the collection into an unusable state, and then fix it by setting website_id. Maybe you can't change a protocol once you create the Collection.  

Or maybe maybe changing the protocol modifies the available fields in the interface, so that you can enter the new settings and then hit submit. Then this method would work fine, but I think you'd need a separate controller to explain what the interface should look like for the new protocol.